File: desktop/src/surface/adapters/review.ts
import type { ReviewContext, ReviewData, ReviewItem, ReviewQuestionBinding, ReviewQuestionReferences, ReviewTarget } from "../types";
import { isRecord, optionalNonNegativeInteger } from "./primitives";

const ACTIONS = ["open_question", "open_transaction"] as const;

function exact(raw: Record<string, unknown>, keys: readonly string[]): boolean {
  const actual = Object.keys(raw).sort();
  const expected = [...keys].sort();
  return actual.length === expected.length && actual.every((key, index) => key === expected[index]);
}

function string(raw: unknown, allowEmpty = false): string | null {
  return typeof raw === "string" && (allowEmpty || raw.trim().length > 0) ? raw : null;
}

function context(raw: unknown): ReviewContext | null {
  if (!isRecord(raw) || !exact(raw, ["date", "amount", "account", "merchant"])) return null;
  const date = string(raw.date, true), amount = string(raw.amount, true);
  const account = string(raw.account, true), merchant = string(raw.merchant, true);
  return date === null || amount === null || account === null || merchant === null
    ? null : { date, amount, account, merchant };
}

export function adaptReviewTarget(raw: unknown): ReviewTarget | null {
  if (!isRecord(raw)) return null;
  const questionId = string(raw.question_id);
  if (!questionId) return null;
  if (raw.kind === "conversation") {
    if (!exact(raw, ["kind", "question_id", "disclosure"])) return null;
    const disclosure = string(raw.disclosure);
    return disclosure ? { kind: "conversation", questionId, disclosure } : null;
  }
  if (raw.kind !== "transaction" || !exact(raw, ["kind", "question_id", "account_id", "movement_id", "canonical_movement_id", "member_movement_ids"])) return null;
  const accountId = string(raw.account_id), movementId = string(raw.movement_id);
  const canonicalMovementId = string(raw.canonical_movement_id);
  if (!Array.isArray(raw.member_movement_ids)) return null;
  const memberMovementIds = raw.member_movement_ids.map((identity) => string(identity));
  if (!accountId || !movementId || !canonicalMovementId || memberMovementIds.some((identity) => identity === null)) return null;
  const safeMembers = memberMovementIds as string[];
  if (!safeMembers.length || new Set(safeMembers).size !== safeMembers.length
      || safeMembers[0] !== canonicalMovementId
      || JSON.stringify(safeMembers) !== JSON.stringify([...safeMembers].sort())
      || !safeMembers.includes(movementId)) return null;
  return { kind: "transaction", questionId, accountId, requestedMovementId: movementId, canonicalMovementId, memberMovementIds: safeMembers };
}

function bindingRefs(raw: unknown): ReviewQuestionReferences | null {
  if (!isRecord(raw) || !exact(raw, ["movement", "movements", "candidates", "document", "doc_id", "account"])) return null;
  const movement = string(raw.movement, true), document = string(raw.document, true);
  const documentId = string(raw.doc_id, true), account = string(raw.account, true);
  if (movement === null || document === null || documentId === null || account === null
      || !Array.isArray(raw.movements) || !Array.isArray(raw.candidates)) return null;
  const movements = raw.movements.map((identity) => string(identity));
  const candidates = raw.candidates.map((identity) => string(identity));
  if (movements.some((identity) => identity === null) || candidates.some((identity) => identity === null)) return null;
  const safeMovements = movements as string[], safeCandidates = candidates as string[];
  if (new Set(safeMovements).size !== safeMovements.length || new Set(safeCandidates).size !== safeCandidates.length) return null;
  return { movement, movements: safeMovements, candidates: safeCandidates, document, documentId, account };
}

export function adaptReviewBinding(raw: unknown): ReviewQuestionBinding | null {
  if (!isRecord(raw) || !exact(raw, ["item_id", "question_id", "question_kind", "label", "reason", "refs", "target", "status", "primary_action", "allowed_actions"])) return null;
  const itemId = string(raw.item_id), questionId = string(raw.question_id);
  const questionKind = string(raw.question_kind), label = string(raw.label), reason = string(raw.reason);
  const refs = bindingRefs(raw.refs), parsedTarget = adaptReviewTarget(raw.target);
  const primaryAction = ACTIONS.find((action) => action === raw.primary_action);
  if (!Array.isArray(raw.allowed_actions)) return null;
  const allowedActions = raw.allowed_actions.map((action) => ACTIONS.find((candidate) => candidate === action));
  if (!itemId || !questionId || !questionKind || !label || !reason || !refs || !parsedTarget
      || raw.status !== "open" || !primaryAction || allowedActions.some((action) => !action)
      || allowedActions.length !== 1 || allowedActions[0] !== primaryAction
      || itemId !== `question:${questionId}` || parsedTarget.questionId !== questionId
      || (primaryAction === "open_transaction") !== (parsedTarget.kind === "transaction")) return null;
  return { itemId, questionId, questionKind, label, reason, refs, target: parsedTarget, status: "open", primaryAction, allowedActions: allowedActions as ReviewQuestionBinding["allowedActions"] };
}

function sameTarget(left: ReviewTarget, right: ReviewTarget): boolean {
  if (left.kind !== right.kind || left.questionId !== right.questionId) return false;
  if (left.kind === "conversation" || right.kind === "conversation") return left.kind === "conversation" && right.kind === "conversation" && left.disclosure === right.disclosure;
  return left.accountId === right.accountId
    && left.requestedMovementId === right.requestedMovementId
    && left.canonicalMovementId === right.canonicalMovementId
    && left.memberMovementIds.length === right.memberMovementIds.length
    && left.memberMovementIds.every((identity, index) => identity === right.memberMovementIds[index]);
}

function item(raw: unknown): ReviewItem | null {
  if (!isRecord(raw) || !exact(raw, ["id", "type", "type_label", "marker", "marker_label", "label", "reason", "status", "context", "target", "primary_action", "action_label", "allowed_actions", "binding"])) return null;
  const id = string(raw.id), label = string(raw.label), reason = string(raw.reason);
  const typeLabel = string(raw.type_label), markerLabel = string(raw.marker_label), actionLabel = string(raw.action_label);
  const parsedContext = context(raw.context), parsedTarget = adaptReviewTarget(raw.target);
  const binding = adaptReviewBinding(raw.binding);
  const primaryAction = ACTIONS.find((action) => action === raw.primary_action);
  if (!Array.isArray(raw.allowed_actions)) return null;
  const allowedActions = raw.allowed_actions.map((action) => ACTIONS.find((candidate) => candidate === action));
  if (!id || !label || !reason || !typeLabel || !markerLabel || !actionLabel || raw.type !== "question" || raw.marker !== "?" || raw.status !== "open"
      || !parsedContext || !parsedTarget || !binding || !primaryAction || allowedActions.some((action) => !action)
      || allowedActions.length !== 1 || allowedActions[0] !== primaryAction
      || (primaryAction === "open_transaction") !== (parsedTarget.kind === "transaction")
      || id !== `question:${parsedTarget.questionId}` || binding.itemId !== id
      || binding.label !== label || binding.reason !== reason || binding.status !== raw.status
      || binding.primaryAction !== primaryAction
      || binding.allowedActions.length !== allowedActions.length
      || !binding.allowedActions.every((action, index) => action === allowedActions[index])
      || !sameTarget(binding.target, parsedTarget)) return null;
  return { id, type: "question", typeLabel, marker: "?", markerLabel, label, reason, status: "open", context: parsedContext, target: parsedTarget, primaryAction, actionLabel, allowedActions: allowedActions as ReviewItem["allowedActions"], binding };
}

export function adaptReview(raw: unknown): ReviewData | null {
  if (!isRecord(raw) || !exact(raw, ["state", "contract", "title", "summary", "actionable_count", "shown_count", "remaining_count", "types", "groups"])
      || raw.state !== "ready" || raw.contract !== "ReviewSummary.v1" || !Array.isArray(raw.types) || !Array.isArray(raw.groups)) return null;
  const actionableCount = optionalNonNegativeInteger(raw.actionable_count);
  const shownCount = optionalNonNegativeInteger(raw.shown_count);
  const remainingCount = optionalNonNegativeInteger(raw.remaining_count);
  if (actionableCount === undefined || shownCount === undefined || remainingCount === undefined || actionableCount !== shownCount + remainingCount) return null;
  const types = raw.types.map((entry) => {
    if (!isRecord(entry) || !exact(entry, ["id", "label", "count"]) || entry.id !== "questions") return null;
    const count = optionalNonNegativeInteger(entry.count), label = string(entry.label);
    return count === undefined || !label ? null : { id: "questions" as const, label, count };
  });
  if (types.some((entry) => entry === null) || new Set(types.map((entry) => entry?.id)).size !== types.length) return null;
  const groups = raw.groups.map((entry) => {
    if (!isRecord(entry) || !exact(entry, ["id", "label", "count", "items"]) || entry.id !== "questions" || !Array.isArray(entry.items)) return null;
    const count = optionalNonNegativeInteger(entry.count), label = string(entry.label);
    const items = entry.items.map(item);
    if (count === undefined || !label || items.some((row) => row === null) || count !== items.length) return null;
    return { id: "questions" as const, label, count, items: items as ReviewItem[] };
  });
  if (groups.some((entry) => entry === null) || new Set(groups.map((entry) => entry?.id)).size !== groups.length) return null;
  const items = groups.flatMap((group) => group?.items ?? []);
  if (items.length !== shownCount || new Set(items.map((row) => row.id)).size !== items.length
      || new Set(items.map((row) => row.target.questionId)).size !== items.length) return null;
  if ((shownCount === 0) !== (groups.length === 0) || (shownCount === 0) !== (types.length === 0)) return null;
  if (shownCount > 0 && (groups.length !== 1 || types.length !== 1
      || types[0]?.count !== shownCount || groups[0]?.count !== shownCount
      || types[0]?.label !== groups[0]?.label)) return null;
  const title = string(raw.title), summary = string(raw.summary);
  if (!title || !summary) return null;
  return { contract: "ReviewSummary.v1", title, summary, actionableCount, shownCount, remainingCount, types: types as ReviewData["types"], groups: groups as ReviewData["groups"] };
}

File: product/viva/surface/review.py
"""Backend-authored attention queue for the Review destination.

Review is deliberately narrower than Activity.  It contains only work another
backend read explicitly says is waiting for the person; editability, an empty
category, or any other shape of a movement is never promoted into attention
here.  Question order is the order supplied by ``open_questions``.
"""

from __future__ import annotations

from typing import Any

from ..questions import ACTIONABLE_QUESTION_WINDOW, open_questions
from .account_ledger import AccountLedgerIdentityError, _deduplicate
from .activity import _row


DEFAULT_LIMIT = ACTIONABLE_QUESTION_WINDOW
MAX_LIMIT = 500
_BINDING_SCALAR_REFS = ("movement", "document", "doc_id", "account")
_BINDING_LIST_REFS = ("movements", "candidates")


def review(projection: Any, locale: str, *, limit: int = DEFAULT_LIMIT,
           as_of: str = "", jurisdiction: str = "") -> dict[str, Any]:
    """Return the ordered, bounded set of authored attention items.

    A transaction destination is included only when every referenced movement
    resolves uniquely, belongs to one account, and collapses to one canonical
    account-ledger row.  Otherwise the item remains actionable through its
    exact question identity and says why transaction context was withheld.
    """
    if isinstance(limit, bool) or not isinstance(limit, int) or not 1 <= limit <= MAX_LIMIT:
        raise ValueError(f"review limit must be between 1 and {MAX_LIMIT}")
    queue = open_questions(
        projection, limit=limit, as_of=as_of,
        jurisdiction=jurisdiction, locale=locale)
    questions = list(queue.get("questions") or [])
    items = [_item(projection, question, locale) for question in questions]
    if any(not item["id"] for item in items) or len({item["id"] for item in items}) != len(items):
        raise ValueError("review item identity is missing or ambiguous")
    total = int(queue.get("total", len(items)))
    if total < len(items):
        raise ValueError("review total is smaller than its supplied items")
    group = {
        "id": "questions",
        "label": "Questions",
        "count": len(items),
        "items": items,
    }
    return {
        "state": "ready",
        "contract": "ReviewSummary.v1",
        "title": "Review",
        "summary": ("Nothing is waiting for your answer."
                    if total == 0 else
                    f"{total} item{' is' if total == 1 else 's are'} waiting for your answer."),
        "actionable_count": total,
        "shown_count": len(items),
        "remaining_count": total - len(items),
        # Questions are the only attention kind this read currently authors.
        # Other kinds join this list only when their own backend projection can
        # supply stable identities, reasons, targets, and actions.
        "types": ([{"id": "questions", "label": "Questions",
                    "count": len(items)}] if items else []),
        "groups": [group] if items else [],
    }


def _item(projection: Any, question: dict[str, Any], locale: str) -> dict[str, Any]:
    question_id = str(question.get("id") or "")
    target, context = _transaction_target(projection, question, locale)
    if target is None:
        target = {
            "kind": "conversation",
            "question_id": question_id,
            "disclosure": "No exact account-ledger transaction could be proven from this question’s references. The question will open in its conversation instead.",
        }
        context = {"date": "", "amount": "", "account": "", "merchant": ""}
        primary_action = "open_question"
        action_label = "Answer question"
    else:
        primary_action = "open_transaction"
        action_label = "Review transaction"
    item = {
        "id": f"question:{question_id}",
        "type": "question",
        "type_label": "Question",
        "marker": "?",
        "marker_label": "Viva needs an answer",
        "label": str(question.get("text") or ""),
        "reason": str(question.get("why") or ""),
        "status": "open",
        "context": context,
        "target": target,
        "primary_action": primary_action,
        "action_label": action_label,
        "allowed_actions": [primary_action],
    }
    item["binding"] = _binding(question, item)
    return item


def question_review_binding(projection: Any, question: dict[str, Any],
                            locale: str) -> dict[str, Any]:
    """Return the exact Review semantics paired with one conversation row."""
    return dict(_item(projection, question, locale)["binding"])


def _binding(question: dict[str, Any], item: dict[str, Any]) -> dict[str, Any]:
    refs = question.get("refs") if isinstance(question.get("refs"), dict) else {}
    normalized_refs: dict[str, Any] = {}
    for key in _BINDING_SCALAR_REFS:
        value = refs.get(key)
        normalized_refs[key] = value if isinstance(value, str) else ""
    for key in _BINDING_LIST_REFS:
        value = refs.get(key)
        normalized_refs[key] = (list(value) if isinstance(value, list)
                                and all(isinstance(entry, str) for entry in value)
                                else [])
    return {
        "item_id": item["id"],
        "question_id": item["target"]["question_id"],
        "question_kind": str(question.get("kind") or ""),
        "label": item["label"],
        "reason": item["reason"],
        "refs": normalized_refs,
        "target": dict(item["target"]),
        "status": item["status"],
        "primary_action": item["primary_action"],
        "allowed_actions": list(item["allowed_actions"]),
    }


def _transaction_target(projection: Any, question: dict[str, Any],
                        locale: str) -> tuple[dict[str, Any] | None, dict[str, str]]:
    refs = question.get("refs")
    if not isinstance(refs, dict):
        return None, {}
    source = refs.get("movement")
    if source not in (None, "") and (not isinstance(source, str) or not source):
        return None, {}
    related = refs.get("movements", [])
    if related is None:
        related = []
    if (not isinstance(related, list)
            or any(not isinstance(value, str) or not value for value in related)
            or len(set(related)) != len(related)):
        return None, {}
    # ``movement`` and ``movements`` are two authored views of the same
    # relationship, so the singular identity may legitimately also occur in
    # the plural list. Repetition *inside* either list is malformed and is
    # refused above rather than silently normalised.
    identities = list(related)
    if isinstance(source, str) and source and source not in identities:
        identities.insert(0, source)
    if not identities:
        return None, {}

    all_movements = list(projection.movements())
    by_id: dict[str, list[Any]] = {}
    for movement in all_movements:
        by_id.setdefault(str(movement.key), []).append(movement)
    if any(len(by_id.get(identity, [])) != 1 for identity in identities):
        return None, {}
    selected = [by_id[identity][0] for identity in identities]
    accounts = {str(movement.account) for movement in selected}
    if len(accounts) != 1:
        return None, {}
    account_id = next(iter(accounts))
    declared_account = refs.get("account")
    if declared_account not in (None, "", account_id):
        return None, {}

    # Candidate references are not destinations, but they are part of the
    # question's claimed relationship.  A malformed or ambiguous candidate
    # makes that relationship unsafe to present as exact transaction context.
    candidates = refs.get("candidates", [])
    if candidates is None:
        candidates = []
    if (not isinstance(candidates, list)
            or any(not isinstance(identity, str) or not identity for identity in candidates)
            or len(set(candidates)) != len(candidates)):
        return None, {}
    if any(len(by_id.get(identity, [])) != 1 for identity in candidates):
        return None, {}

    infos = [info for info in projection.account_infos()
             if str(info.account) == account_id]
    if len(infos) != 1:
        return None, {}
    statements = projection.statements(account_id)
    records = list(statements.records) if statements else []
    try:
        entries, _ = _deduplicate(
            [movement for movement in all_movements
             if str(movement.account) == account_id], records)
    except AccountLedgerIdentityError:
        return None, {}
    containing = [entry for entry in entries if identities and all(
        identity in {str(member.key) for member in entry["members"]}
        for identity in identities)]
    if len(containing) != 1:
        return None, {}
    entry = containing[0]
    member_ids = sorted(str(member.key) for member in entry["members"])
    canonical = str(entry["movement"].key)
    requested = str(source or identities[0])
    if requested not in member_ids or canonical not in member_ids:
        return None, {}

    # Candidates are a claim about this question's movement relationship, not
    # free-floating context. A candidate outside the one proven canonical row
    # would make opening that row imply a relationship the projection did not
    # establish.
    if any(identity not in member_ids for identity in candidates):
        return None, {}

    document_alias = refs.get("document")
    document_id_alias = refs.get("doc_id")
    if (document_alias not in (None, "")
            and (not isinstance(document_alias, str) or not document_alias)):
        return None, {}
    if (document_id_alias not in (None, "")
            and (not isinstance(document_id_alias, str) or not document_id_alias)):
        return None, {}
    if (document_alias and document_id_alias and document_alias != document_id_alias):
        return None, {}
    document = document_alias or document_id_alias
    if document:
        member_documents = {
            str(getattr(member.provenance, "doc_id", "") or "")
            for member in entry["members"]}
        if document not in member_documents:
            return None, {}

    row = _row(projection, entry["movement"], locale, {
        "categories": {"items": [], "complete": False, "limit": 0},
        "subcategories": {"items": [], "complete": False, "limit": 0},
        "tags": {"items": [], "complete": False, "limit": 0,
                 "max_selected": 0, "max_label_length": 0},
    })
    return ({
        "kind": "transaction",
        "question_id": str(question.get("id") or ""),
        "account_id": account_id,
        "movement_id": requested,
        "canonical_movement_id": canonical,
        "member_movement_ids": member_ids,
    }, {
        "date": str(row.get("date") or ""),
        "amount": str(row.get("display") or ""),
        "account": str(row.get("account_name") or ""),
        "merchant": str(row.get("description") or ""),
    })
